models/spring-ai-ollama/src/main/java/org/springframework/ai/ollama/OllamaChatModel.java

```diff
@@ -271,10 +272,19 @@ private ChatResponse internalCall(Prompt prompt, ChatResponse previousChatRespon
 
 		if (ToolCallingChatOptions.isInternalToolExecutionEnabled(prompt.getOptions()) && response != null
 				&& response.hasToolCalls()) {
-			var toolCallConversation = this.toolCallingManager.executeToolCalls(prompt, response);
-			// Recursively call the call method with the tool call message
-			// conversation that contains the call responses.
-			return this.internalCall(new Prompt(toolCallConversation, prompt.getOptions()), response);
+			var toolExecutionResult = this.toolCallingManager.executeToolCalls(prompt, response);
+			if (toolExecutionResult.returnDirect()) {
+				// Return tool execution result directly to the client.
+				return ChatResponse.builder()
```

Comment:
When a given tool is marked with the option `returnDirect`, the tool call result is sent back to the user directly, wrapping each tool call result in a `Generation` object.

That is especially useful when building agents. For example, if I have a RAG tool, I might want to get the result back directly instead of having the main agent model post-process the RAG answer. Or I might want certain tools to end the reasoning loop of an agent.